File: 2D Elastic FDTD/teflon_fibre_elastic.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <complex.h>

#define NX 500
#define NZ 500
#define MAX_ITER 10000

// Memory-safe 2D mapping (Row-Major to match C)
#define TXX(x,z)    tau_xx[(x)*NZ + (z)]
#define TZZ(x,z)    tau_zz[(x)*NZ + (z)]
#define TXZ(x,z)    tau_xz[(x)*(NZ-1) + (z)]
#define UX(x,z)     ux[(x)*NZ + (z)]
#define UZ(x,z)     uz[(x)*(NZ-1) + (z)]
#define TXX_MOD(x,z) tau_xx_mod[(x)*NZ + (z)]
#define UZ_MOD(x,z)  uz_mod[(x)*(NZ-1) + (z)]
#define UX_MOD(x,z)  ux_mod[(x)*NZ + (z)]

int main() {
    int xind, zind, n;
    float pi = 3.14159265f, freq = 1.7e6f, cmax = 1600.0f;
    float dx = 5e-6f, dz = 5e-6f;
    float dt = 0.95f * (dx / (sqrtf(2.0f) * cmax));
    
    // Allocate all arrays on heap to avoid stack overflow
    float *l = calloc(NX*NZ, sizeof(float));
    float *m = calloc(NX*NZ, sizeof(float));
    float *rho = calloc(NX*NZ, sizeof(float));
    float *b = calloc(NX*NZ, sizeof(float));
    float *tau_xx = calloc(NX*NZ, sizeof(float));
    float *tau_zz = calloc(NX*NZ, sizeof(float));
    float *tau_xx_n = calloc(NX*NZ, sizeof(float));
    float *tau_zz_n = calloc(NX*NZ, sizeof(float));
    float *tau_xz = calloc((NX-1)*(NZ-1), sizeof(float));
    float *ux = calloc((NX-1)*NZ, sizeof(float));
    float *uz = calloc(NX*(NZ-1), sizeof(float));
    
    float complex *ux_mod = calloc((NX-1)*NZ, sizeof(float complex));
    float complex *uz_mod = calloc(NX*(NZ-1), sizeof(float complex));

    int center_x = NX/2;
    int center_z = NZ/2;

    // 1. Initialize Background (Water)
    for(int i=0; i<NX*NZ; i++) {
        rho[i] = 1000.0f;
        l[i] = 1000.0f * powf(1540.0f, 2.0f);
    }

    // 2. Define Waveguide Structure (Circle)
    for(xind=0; xind<NX; xind++) {
        for(zind=0; zind<NZ; zind++) {
            float dist = sqrtf(powf(center_x-xind, 2) + powf(center_z-zind, 2));
            if(dist < (525e-6f/dx/2.0f)) {
                l[xind*NZ+zind] = 1100.0f * (powf(1400.0f, 2.0f) - 2.0f*powf(440.0f, 2.0f));
                m[xind*NZ+zind] = 1100.0f * powf(440.0f, 2.0f);
                rho[xind*NZ+zind] = 1100.0f;
            }
            if(dist < (250e-6f/dx/2.0f)) {
                l[xind*NZ+zind] = 1000.0f * powf(1540.0f, 2.0f);
                m[xind*NZ+zind] = 0.0f;
                rho[xind*NZ+zind] = 1000.0f;
            }
            b[xind*NZ+zind] = 1.0f / rho[xind*NZ+zind];
        }
    }

    printf("Starting FDTD loop. dt = %e\n", dt);

    // 3. Main FDTD Loop
    for(n=0; n <= MAX_ITER; n++) {
        
        // Update Velocities (UX, UZ)
        for(xind=1; xind<NX-1; xind++) {
            for(zind=1; zind<NZ-1; zind++) {
                UX(xind,zind) += dt * b[xind*NZ+zind] * ((TXX(xind,zind) - TXX(xind-1,zind))/dx + (TXZ(xind,zind) - TXZ(xind,zind-1))/dz);
                UZ(xind,zind) += dt * b[xind*NZ+zind] * ((TXZ(xind,zind) - TXZ(xind-1,zind))/dx + (TZZ(xind,zind) - TZZ(xind,zind-1))/dz);
            }
        }

        // Update Stresses (TXX, TZZ)
        for(xind=0; xind<NX-2; xind++) {
            for(zind=0; zind<NZ-2; zind++) {
                float dux = (UX(xind+1,zind) - UX(xind,zind))/dx;
                float duz = (UZ(xind,zind+1) - UZ(xind,zind))/dz;
                TXX(xind,zind) += dt * (l[xind*NZ+zind] + 2.0f*m[xind*NZ+zind]) * dux + dt * l[xind*NZ+zind] * duz;
                TZZ(xind,zind) += dt * (l[xind*NZ+zind] + 2.0f*m[xind*NZ+zind]) * duz + dt * l[xind*NZ+zind] * dux;
            }
        }

        // Update Shear Stress (TXZ)
        for(xind=0; xind<NX-1; xind++) {
            for(zind=0; zind<NZ-1; zind++) {
                TXZ(xind,zind) += dt * m[xind*NZ+zind] * ((UX(xind,zind+1) - UX(xind,zind))/dz + (UZ(xind+1,zind) - UZ(xind,zind))/dz);
            }
        }

        // Source Injection (Moved to center-top)
        TXX(center_x, 5) += sinf(2.0f * pi * freq * n * dt);
        TZZ(center_x, 5) += sinf(2.0f * pi * freq * n * dt);

        // Mur1 ABC (Note: not all edges of the computational domain are implemented)
        float c_abs = 1540.0f;
        float factor = (c_abs*dt - dx) / (c_abs*dt + dx);
        for(zind=0; zind<NZ; zind++) {
            TXX(0, zind) = tau_xx_n[1*NZ+zind] + factor*(TXX(1,zind) - tau_xx_n[0*NZ+zind]);
            TXX(NX-1, zind) = tau_xx_n[(NX-2)*NZ+zind] + factor*(TXX(NX-2,zind) - tau_xx_n[(NX-1)*NZ+zind]);
        }

        // Store history for ABC
        for(int i=0; i<NX*NZ; i++) { tau_xx_n[i] = tau_xx[i]; tau_zz_n[i] = tau_zz[i]; }

        // Accumulate Complex Fields for FFT/Phasor
        float complex phase = cexpf(I * n * dt * 2.0f * pi * freq);
        for(xind=0; xind<NX-1; xind++) {
            for(zind=0; zind<NZ; zind++) UX_MOD(xind,zind) += UX(xind,zind) * phase;
        }
        for(xind=0; xind<NX; xind++) {
            for(zind=0; zind<NZ-1; zind++) UZ_MOD(xind,zind) += UZ(xind,zind) * phase;
        }

        if (n % 100 == 0) {
            printf("Step %d | Time: %.2f us\n", n, (n*dt)/1e-6);
            FILE *f = fopen("abs_u.out", "w");
            for(zind=0; zind<NZ-1; zind++) {
                for(xind=0; xind<NX-1; xind++) {
                    float val = cabsf(UX_MOD(xind,zind)) + cabsf(UZ_MOD(xind,zind));
                    fprintf(f, "%e ", val);
                }
                fprintf(f, "\n");
            }
            fclose(f);
        }
    }

    return 0;
}

File: 2D Elastic FDTD/teflon_fibre_elastic.f90
	program fdtd

	implicit none

    integer nx, nz, center_x, center_z
    parameter (nx = 500, nz = 500)
    integer xind, zind, i, k, n
    integer max_iter
    parameter (max_iter = 10000)
    
    real*8 l(nx,nz), m(nx,nz)
    real*8 rho(nx,nz), b(nx,nz)
    real*8 c, cmax
    real*8 t0, width, freq, omega, pi, lambda
    real*8 dx, dz, dt, tw
    real*8 tau_xx_det(max_iter), tau_zz_det(max_iter), u_det(max_iter)
    
    real*8 tau_xx(nx,nz)       
    real*8 tau_zz(nx,nz)
    real*8 tau_xz(nx-1,nz-1)
    real*8 ux(nx-1,nz)    
    real*8 uz(nx,nz-1)    

    real*8 tau_xx_n(nx,nz)       
    real*8 tau_zz_n(nx,nz)       
    real*8 ux_n(nx-1,nz)    
    real*8 uz_n(nx,nz-1)    

    complex*16 tau_xx_mod(nx,nz)
    complex*16 tau_zz_mod(nx,nz)       
    complex*16 tau_xz_mod(nx-1,nz-1)       
    complex*16 ux_mod(nx-1,nz)    
    complex*16 uz_mod(nx,nz-1)    

    complex*16 COMPLEX_I    
    real*8 tbplt(nx-1,nz-1), div_x, div_z, xdist, zdist, dist
    character*32 filename
   
    tau_xx = 0.0       
    tau_zz = 0.0
    tau_xz = 0.0
    ux = 0.0    
    uz = 0.0    
   
    COMPLEX_I = dcmplx(0.0, 1.0)

    t0 = 500.0       !index time of gaussian pulse peak
    width = 1000.0     !peakyness of gaussian pulse

    pi = 3.1415926
    freq = 1.7e6
    cmax = 1600.0
    omega = 2*pi*freq
    lambda = cmax/freq

    dx = 5e-6
    dz = dx

    !! Initialise the p and u fields and set coefficients + a Courrant compliant dt
        
    dt = dx/(dsqrt(2.0D0)*cmax) !set time step small enough to satisfy courrant condition
    dt = 0.95*dt
    
    write(*,*) 'dt = ', dt

    !definition of the waveguide structure
    center_x = int(nx/2)-1
    center_z = int(nz/2)-1
    
    rho = 1000.0
    l = rho*(1540.0**2)
    m = 0.0       

    do zind = 1, nz
    do xind = 1, nx
     
        zdist = (center_z-zind)
        xdist = (center_x-xind)
        dist = dsqrt(xdist**2+zdist**2)
        
        if(dist<525e-6/dx/2.0) then
        
		l(xind,zind) = 1100.0*(1400.0**2 - 2.0*440.0**2)
		m(xind,zind) = 1100.0*440.0**2       			
		rho(xind,zind) = 1100.0
    
        endif
        
    enddo
    enddo
    
    do zind = 1, nz
    do xind = 1, nx
     
        zdist = (center_z-zind)
        xdist = (center_x-xind)
        dist = dsqrt(xdist**2+zdist**2)
        
        if(dist<250e-6/dx/2.0) then
        
		l(xind,zind) = 1000.0*1540.0**2
		m(xind,zind) = 0.0
		rho(xind,zind) = 1000.0
    
        endif
        
    enddo
    enddo

	b = 1.0/rho 
       
    !! FDTD loop - comment out plot lines for speed
    do n = 0,max_iter
    
    ! Update the ux values:
    do zind=2,nz-1
    do xind=2,nx-1
    
    ux(xind,zind) = ux(xind,zind) &
    + dt*b(xind,zind)*((tau_xx(xind,zind) - tau_xx(xind-1,zind))/dx &
    + (tau_xz(xind,zind) - tau_xz(xind,zind-1))/dz)
    enddo
    enddo

    do zind=2,nz-1
    do xind=2,nx-1

    uz(xind,zind) = uz(xind,zind) &
    + dt*b(xind,zind)*((tau_xz(xind,zind) - tau_xz(xind-1,zind))/dx &
    + (tau_zz(xind,zind) - tau_zz(xind,zind-1))/dz)
 
    enddo
    enddo

    do zind=1,nz-2
    do xind=1,nx-2
    
    tau_xx(xind,zind) = tau_xx(xind,zind) + dt*(l(xind,zind)+2.0*m(xind,zind))*((ux(xind+1,zind) - ux(xind,zind))/dx) + &
    dt*l(xind,zind)*(uz(xind,zind+1) - uz(xind,zind))/dz

    tau_zz(xind,zind) = tau_zz(xind,zind) + dt*(l(xind,zind)+2.0*m(xind,zind))*((uz(xind,zind+1) - uz(xind,zind))/dz) + &
    dt*l(xind,zind)*(ux(xind+1,zind) - ux(xind,zind))/dx

    enddo
    enddo
    
    do zind=1,nz-1
    do xind=1,nx-1
    
    tau_xz(xind,zind) = tau_xz(xind,zind) + dt*m(xind,zind)*((ux(xind,zind+1) - ux(xind,zind))/dz &
    + (uz(xind+1,zind) - uz(xind,zind))/dz)

    enddo
    enddo

    tau_xx(center_x,5) = tau_xx(center_x,5) + dsin(2.0*pi*n*dt*freq) 
    tau_zz(center_x,5) = tau_zz(center_x,5) + dsin(2.0*pi*n*dt*freq) 

        
    !Mur1 ABC 

    c = 1540.0
   

	do zind=1,nz
    do xind=1,nx

	!LEFT WALL:
        tau_xx(1,zind)=tau_xx_n(2,zind)+((c*dt-dx)/(c*dt+dx))*(tau_xx(2,zind)-tau_xx_n(1,zind))
        tau_zz(1,zind)=tau_zz_n(2,zind)+((c*dt-dx)/(c*dt+dx))*(tau_zz(2,zind)-tau_zz_n(1,zind))

        !RIGHT WALL:
        tau_xx(nx-1,zind)=tau_xx_n(nx-1-1,zind)+((c*dt-dx)/(c*dt+dx))*(tau_xx(nx-1-1,zind)-tau_xx_n(nx-1,zind))
        tau_zz(nx-1,zind)=tau_zz_n(nx-1-1,zind)+((c*dt-dx)/(c*dt+dx))*(tau_zz(nx-1-1,zind)-tau_zz_n(nx-1,zind))

        !TOP WALL
        tau_xx(xind,1)=tau_xx_n(xind,2)+((c*dt-dz)/(c*dt+dz))*(tau_xx(xind,2)-tau_xx_n(xind,1))
        tau_zz(xind,1)=tau_zz_n(xind,2)+((c*dt-dz)/(c*dt+dz))*(tau_zz(xind,2)-tau_zz_n(xind,1))

        !BOTTOM WALL
        tau_xx(xind,nz-1)=tau_xx_n(xind,nz-1-1)+((c*dt-dz)/(c*dt+dz))*(tau_xx(xind,nz-1-1)-tau_xx_n(xind,nz-1))
        tau_zz(xind,nz-1)=tau_zz_n(xind,nz-1-1)+((c*dt-dz)/(c*dt+dz))*(tau_zz(xind,nz-1-1)-tau_zz_n(xind,nz-1))       

    enddo
    enddo

    !save 'old' fields for Mur ABC
    ux_n=ux
    uz_n=uz
    tau_xx_n = tau_xx
    tau_zz_n = tau_zz
    
    !symmetry
    ux(center_x:nx-1,1:nz) = 0.0
    uz(center_x:nx,1:nz-1) = 0.0
    tau_xx(center_x:nx,1:nz) = 0.0
    tau_zz(center_x:nx,1:nz) = 0.0  
    tau_xz(center_x:nx-1,1:nz-1) = 0.0      
              
    if ( mod(n,int(0.1e-6/dt)).eq.0) then   
       
        tau_xx_mod = tau_xx_mod + tau_xx*exp(COMPLEX_I*n*dt*2.0*pi*freq)
		tau_zz_mod = tau_zz_mod + tau_zz*exp(COMPLEX_I*n*dt*2.0*pi*freq)
		ux_mod = ux_mod + ux*exp(COMPLEX_I*n*dt*2.0*pi*freq)
		uz_mod = uz_mod + uz*exp(COMPLEX_I*n*dt*2.0*pi*freq)
        
         
    endif

	if ( mod(n,100).eq.0) then   

        write(*,*) 'step #', n, 'time ', dble(n)*dt/1e-6, 'micro sec'
                
        tbplt = dsqrt(abs(ux_mod(1:nx-1,1:nz-1))**2+abs(uz_mod(1:nx-1,1:nz-1))**2)
                
        open (unit=34,file='abs_u.out',status='unknown')
	    do k=1, nz-1
	    do i=1, nx-1
	    
            WRITE(34,"(1X, E18.8E3,$)")  tbplt(i,k)
        ENDDO
	    WRITE(34,*)
        ENDDO
        CLOSE(34)
                 
    endif

    enddo! end FDTD body

    END
